model/gen-ai/events.yaml

```diff
@@ -46,3 +180,111 @@ groups:
+          type: enum
+          members:
+            - id: stop
+              value: 'stop'
+              brief: Stop
+            - id: tool_calls
+              value: 'tool_calls'
+              brief: Tool Calls
+            - id: content_filter
+              value: 'content_filter'
+              brief: Content Filter
+            - id: length
+              value: 'length'
+              brief: Length
+            - id: error
+              value: 'error'
+              brief: Error
+          stability: experimental
+          brief: >
+            The reason the model stopped generating tokens.
+          requirement_level: required
+        - id: message
+          type: map
+          stability: experimental
+          brief: >
+            GenAI response message.
+          requirement_level: recommended
+          fields:
+            - id: content
+              type: undefined
+              stability: experimental
+              brief: >
+                The contents of the assistant message.
+              examples: ["The weather in Paris is rainy and overcast, with temperatures around 57°F"]
+              requirement_level: opt_in
+            - id: role
+              type: string
+              stability: experimental
+              brief: >
+                The actual role of the message author as passed in the message.
+              examples: ["assistant", "bot"]
+              requirement_level:
+                conditionally_required: if available and not equal to `assistant`.
+        - id: tool_calls
+          type: map  # TODO: it's an array
+          stability: experimental
+          brief: >
+            The tool calls generated by the model, such as function calls.
+          requirement_level:
+            conditionally_required: if available
+          fields:
+            - id: id
+              type: string
+              stability: experimental
+              brief: >
+                The id of the tool call.
+              examples: ["call_mszuSIzqtI65i1wAUOE8w5H4"]
+              requirement_level: required
+            - id: type
+              type: enum
+              members:
+                - id: function
+                  value: 'function'
+                  brief: Function
+              stability: experimental
+              brief: >
+                The type of the tool.
+              requirement_level: required
+            - id: function
+              type: map
+              stability: experimental
+              brief: >
+                The function that the model called.
+              requirement_level: required
+              fields:
+                - id: name
+                  type: string
+                  stability: experimental
+                  brief: >
+                    The name of the function.
+                  examples: ["get_weather"]
+                  requirement_level: required
+                - id: arguments
+                  type: undefined
```

Comment:
let's clarify if these should be deserialized or provided as string. 
The perf cost of deserialization compared to all other things necessary to create this event is small.

Reply:
A concern: In general the "arguments" string returned by (at least OpenAI -- my experience is limited) may not be valid JSON. In the (likely rare) case that it isn't this means that "arguments" would then, presumably, fallback to be a string. That could be a rare surprise (i.e. possible bug) to whatever system is processing this telemetry.

I suppose that is fine if the type allows both string and object (or perhaps `AnyValue` would be used) -- then implementations need to cope.

---

Another possible concern: https://platform.openai.com/docs/api-reference/chat/create#chat-create-response_format includes the (surprising to me) warning:

> Important: when using JSON mode, you must also instruct the model to produce JSON yourself via a system or user message. Without this, the model may generate an unending stream of whitespace until the generation reaches the token limit, resulting in a long-running and seemingly "stuck" request.

Not having the raw "arguments" string in the telemetry could make it hard to debug why a parsed JSON response of, say `{"foo": "bar"}`, took 1024 (or however many) tokens.

---

What are reasonable use cases for having the parsed arguments in telemetry?

Reply:
if max token is reached in the middle of the arguments, it's truncated, so let's keep it a string, but we might change it for some systems.

Reply:
Then we should also update the examples above

Reply:
Thank you both, added a note and updated examples.